File: Solutions/Python/evaluate_division.py
"""
    [MEDIUM]
    399. Evaluate Division

    Concepts:
    - graph
    - depth-first search

    Stats:
        Runtime | 2 ms      [Beats 10.31%]
        Memory  | 17.94 MB  [Beats 48.64%]
"""

class Solution:
    def calcEquation(
        self,
        equations: List[List[str]],
        values: List[float],
        queries: List[List[str]]
    ) -> List[float]:

        res = [-1.0]*len(queries)
        # store nodes and their edge weights
        nodes = defaultdict(set)
        edges = defaultdict(set)

        for i, (num, denom) in enumerate(equations):
            # set edges to self
            nodes[num].add(num)
            nodes[denom].add(denom)
            edges[(num, num)] = 1.0 if num else 0.0
            edges[(denom, denom)] = 1.0 if denom else 0.0
            # set edges in both directions
            nodes[num].add(denom)
            nodes[denom].add(num)
            edges[(num, denom)] = values[i] if denom else float("inf")
            edges[(denom, num)] = 1.0 / values[i] if num else float("inf")

        def dfs(
            node: str,
            target: str,
            value: float,
            visited: set[str]
        ) -> float:

            nonlocal val

            if target in nodes[node]:
                val = value * edges[(node, target)]
                return True

            for n in nodes[node]:
                if (
                    n not in visited and
                    dfs(
                        n,
                        target,
                        value * edges[(node, n)],
                        visited | {node}
                    )
                ):
                    return True

            return False

        for i, (num, denom) in enumerate(queries):
            val = None
            dfs(num, denom, 1.0, set())
            if val is not None:
                res[i] = val

        return res

File: Solutions/Cpp/evaluate_division.cpp
/*
    [MEDIUM]
    399. Evaluate Division

    Concepts:
    - graph
    - depth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 12.00 MB  [Beats 53.99%]
*/

class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        const double INF = std::numeric_limits<double>::infinity();
        double val;
        std::vector<double> res(queries.size(), -1.0);
        
        // store nodes and their edge weights
        std::unordered_map<std::string, std::unordered_set<std::string>> nodes = {};
        std::unordered_map<std::string, double> edges = {};
        std::unordered_set<std::string> visited = {};

        std::string num, denom;
        for (int i = 0; i < equations.size(); i++) {
            num = equations[i].front();
            denom = equations[i].back();

            // set edges to self
            nodes[num].insert(num);
            nodes[denom].insert(denom);
            edges[num + "," + num] = 1.0;
            edges[denom + "," + denom] = 1.0;
            // set edges in both directions
            nodes[num].insert(denom);
            nodes[denom].insert(num);
            edges[num + "," + denom] = values[i];
            edges[denom + "," + num] = 1.0 / values[i];
        }

        std::function<bool(std::string, std::string, double)> dfs;
        dfs = [&](std::string node, std::string target, double value) {
            if (!nodes.contains(node))
                return false;

            if (nodes.at(node).contains(target)) {
                val = value * edges.at(node + "," + target);
                return true;
            }

            for (std::string n : nodes.at(node)) {
                if (!visited.contains(n)) {
                    visited.insert(n);
                    if (dfs(n, target, value * edges.at(node + "," + n)))
                        return true;
                    visited.erase(n);
                }
            }

            return false;
        };

        for (int i = 0; i < queries.size(); i++) {
            num = queries[i].front();
            denom = queries[i].back();

            // reset for next iteration
            val = INF;
            visited.clear();
            visited.insert(num);

            dfs(num, denom, 1.0);
            if (val != INF)
                res[i] = val;
        }

        return res;
    }
};